pallets/corporate-actions/src/ballot.rs

```diff
@@ -0,0 +1,521 @@
+
+        /// A vote was cast in a corporate ballot.
+        ///
+        /// (voter DID, CAId, Votes)
+        VoteCast(IdentityId, CAId, Vec<(Balance, u16)>),
+
+        /// A corporate ballot changed its start/end date range.
+        ///
+        /// (Ticker's CAA, CA's ID, Voting start/end)
+        RangeChanged(IdentityId, CAId, BallotRange),
+
+        /// A corporate ballot changed its metadata.
+        ///
+        /// (Ticker's CAA, CA's ID, New metadata)
+        MetaChanged(IdentityId, CAId, BallotMeta),
+
+        /// A corporate ballot changed its RCV support.
+        ///
+        /// (Ticker's CAA, CA's ID, New support)
+        RCVChanged(IdentityId, CAId, bool),
+
+        /// A corporate ballot was removed.
+        ///
+        /// (Ticker's CAA, CA's ID)
+        Removed(IdentityId, CAId),
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// A corporate ballot was made for a non `IssuerNotice` CA.
+        CANotNotice,
+        /// A corporate ballot already exists for this CA.
+        AlreadyExists,
+        /// A corporate ballot doesn't exist for this CA.
+        NoSuchBallot,
+        /// A corporate ballot's start time was strictly after the ballot's end.
+        StartAfterEnd,
+        /// A corporate ballot's end time was strictly before the current time.
+        NowAfterEnd,
+        /// If some motion in a corporate ballot has more choices than would fit in `u16`.
+        NumberOfChoicesOverflow,
+        /// Voting started already. Amending a ballot is no longer possible.
+        VotingAlreadyStarted,
+        /// Voting hasn't started yet.
+        VotingNotStarted,
+        /// Voting ended already.
+        VotingAlreadyEnded,
+        /// Provided list of balances does not match the total number of choices.
+        WrongVoteCount,
+        /// Voting power used by a DID on a motion exceeds that which is available to them.
+        InsufficientVotes,
+        /// The RCV fallback of some choice does not exist.
+        NoSuchRCVFallback,
+        /// RCV is not allowed for this ballot.
+        RCVNotAllowed
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // Compute number-of-choices-in-motion cache for `motions`.
+    fn derive_motion_num_choices(motions: &[Motion]) -> Result<Vec<u16>, DispatchError> {
+        let mut total: usize = 0;
+        motions
+            .iter()
+            .map(|motion| {
+                let len = motion.choices.len();
+                // Overflowing usize here will never happen in practice,
+                // but can happen in theory.
+                // We do this now to avoid the potential overflow in `vote`.
+                total = total.checked_add(len)?;
+                len.try_into().ok()
+            })
+            .collect::<Option<_>>()
+            .ok_or_else(|| Error::<T>::NumberOfChoicesOverflow.into())
+    }
+
+    /// Ensure that `now < range.start`.
+    fn ensure_ballot_unstarted(range: BallotRange) -> DispatchResult {
```

Comment:
"Unstarted" might mean that the ballot's start was interrupted. I think `ensure_ballot_not_started` would sound better.

Reply:
(From the other PR, I'll fix it there.)